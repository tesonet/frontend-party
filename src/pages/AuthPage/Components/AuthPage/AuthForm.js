import React, { Component } from 'react'
import logo from '../../../../assets/images/logo.png'

class AuthForm extends Component {

    state = {
        username: '',
        password: '',
        loginError: false
    }

    static getDerivedStateFromProps(nextProps) {
        return {
            loginError: nextProps.loginError
        }
    }

    componentDidUpdate() {
        if (this.props.withAuth) {
            this.props.history.push('servers')
        }
    }

    onInputChange = e => {
        this.setState({
            [e.target.name]: e.target.value
        })
    }

    onInputFocus = () => {
        if (this.state.loginError) {
            this.props.clearLoginError()
        }
    }

    formSubmit = e => {
        e.preventDefault()
        if (this.state.loginError) {
            this.props.clearLoginError()
        }
        this.props.login(this.state)
    }

    render() {

        const {
            username,
            password,
            loginError
        } = this.state

        const isInvalid =
          password === '' ||
          username === ''

        return (
          <div
            className={`${loginError ? 'error' : ''} auth-form`}
          >
              <img
                className="logo"
                src={logo}
                alt="logo"
              />
              <form
                className="auth-form_form"
                onSubmit={this.formSubmit}
              >
                  <input
                    onFocus={this.onInputFocus}
                    className="auth-form_input user-inp"
                    onChange={this.onInputChange}
                    placeholder="Username"
                    name="username"
                    value={username}
                    type="text"
                  />
                  <input
                    onFocus={this.onInputFocus}
                    className="auth-form_input pass-inp"
                    onChange={this.onInputChange}
                    placeholder="Password"
                    name="password"
                    type="text"
                    value={password}
                  />
                  <button
                    className="auth-form_btn btn"
                    type="submit"
                    disabled={isInvalid}
                  >
                      Log In
                  </button>
              </form>
          </div>
        )
    }
}

export default AuthForm