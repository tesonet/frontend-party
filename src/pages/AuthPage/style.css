
.auth-page, .not-found {
    width: 100%;
    height: 100vh;
    display: flex;
    justify-content: center;
    align-items: center;
}

.not-found_content {
    color: #efefef;
    position: relative;
    text-align: center;
}

.not-found_content_message {
    font-size: 24px;
}

.not-found_content_status {
    font-weight: 600;
    font-size: 60px;
}

.auth-page_bg, .auth-page_overlay, .not-found_overlay, .not-found_bg {
    position: absolute;
    width: 100%;
    height: 100%;
}

.auth-page_bg, .not-found_bg {
    background: url("../../assets/images/main-bg.jpg") no-repeat center;
    background-size: cover;
}

.auth-page_overlay, .not-found_overlay {
    background: #0b0f27;
    opacity: 0.7;
}

.auth-form {
    background: transparent;
    display: flex;
    flex-direction: column;
    align-items: flex-end;
    position: relative;
    max-width: 400px;
    padding: 20px;
}

.logo {
    width: 55%;
    text-align: right;
}

.auth-form_form {
    width: 100%;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: flex-end;
    padding-top: 40px;
}

.auth-form_btn, .auth-form_input {
    width: 100%;
    margin: 5px 0px;
    height: 50px;
    border-radius: 6px;
}

.auth-form_btn {
    background: #9fd533;
    outline-color: #08d51a;
    color: white;
    font-weight: bold;
}

.auth-form_input {
    background: white;
    display: inline-block;
    border: 1px solid white;
    padding: 0px 40px;
    color: gray;
    transition: all .2s ease-in-out;
}

.auth-form.error {
    animation: bounce .8s ease-in-out both;
}

.auth-form.error .auth-form_input {
    border-color: #740000;
}

@keyframes bounce {
    0%, 100%, 20%, 50%, 80% {
        transform:         translateX(0)
    }
    40% {
        transform:         translateX(-30px)
    }
    60% {
        transform:         translateX(-15px)
    }
}