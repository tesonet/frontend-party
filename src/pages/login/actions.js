import { history } from '../../App';
import * as actionTypes from './actionTypes';
import { getUserToken } from '../../server/api/tesonetApi';
import errorMessages from './loginForm/errorMessages';

export const token = (authToken) => ({
  type: actionTypes.SET_TOKEN,
  payload: { authToken },
});

export const logout = () => ({
  type: actionTypes.LOGOUT,
});

export const loading = () => ({
  type: actionTypes.LOADING,
});

export const setError = (message) => ({
  type: actionTypes.SET_ERROR,
  payload: { message },
});

export const login = (values) => async (dispatch) => {
  dispatch(loading());
  const response = await getUserToken(values);
  if (response && response.ok) {
    const data = await response.json();
    localStorage.setItem('token', data.token);
    dispatch(token(data.token));
    history.push('/servers');
  } else if (response && response.status === 401) {
    dispatch(setError(errorMessages.WRONG_CREDENTIALS));
  } else {
    dispatch(setError(errorMessages.ERROR));
  }
};
