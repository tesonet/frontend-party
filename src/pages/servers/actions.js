import { getServersList } from '../../server/api/tesonetApi';
import * as actionTypes from './actionTypes';

const fetchSuccess = (servers) => ({
  type: actionTypes.FETCH_SERVERS,
  payload: { servers },
});

const loading = () => ({
  type: actionTypes.LOADING,
});

const setError = (error) => ({
  type: actionTypes.FETCH_ERROR,
  payload: { error },
});

const fetchServers = (token) => async (dispatch) => {
  const response = await getServersList(token);
  dispatch(loading());
  if (response && response.ok) {
    const data = await response.json();
    dispatch(fetchSuccess(data));
  } else {
    dispatch(setError());
  }
};

export {
  fetchServers,
  loading,
  setError,
};
