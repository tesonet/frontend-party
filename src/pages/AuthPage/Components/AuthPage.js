import React, { Component } from 'react'
import AuthForm from './AuthPage/AuthForm'

class AuthPage extends Component {

    componentDidMount() {
        if (this.props.withAuth) {
            this.props.redirectToAuthHomePage()
        }
    }

    render() {
        const { login, loginError, clearLoginError } = this.props

        return <div className="auth-page">
            <div className="auth-page_bg"/>
            <div className="auth-page_overlay"/>
            <AuthForm
              loginError={loginError}
              login={login}
              clearLoginError={clearLoginError}
            />
        </div>
    }
}

export default AuthPage