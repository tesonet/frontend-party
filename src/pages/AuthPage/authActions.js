import axios from 'axios'
import { ASYNC_ACTION_END, ASYNC_ACTION_START } from '../../constants/actions'

export const AUTHORIZE = '[AUTH]AUTHORIZE'
export const LOGIN_START = '[AUTH]LOGIN_START'
export const LOGIN_SUCCESS = '[AUTH]LOGIN_SUCCESS'
export const LOGIN_ERROR = '[AUTH]LOGIN_ERROR'
export const CLEAR_LOGIN_ERROR = '[AUTH]CLEAR_LOGIN_ERROR'
export const LOGOUT = '[AUTH]LOGOUT'

const TOKEN_URL = 'http://playground.tesonet.lt/v1/tokens'

const login = (user) => (
  async (dispatch) => {
      try {
          dispatch({ type: ASYNC_ACTION_START })
          dispatch({ type: LOGIN_START })

          const config = {
              method: 'post',
              url: TOKEN_URL,
              data: user
          }

          const response = await axios(config)

          if (response.status === 200) {
              dispatch({
                  type: LOGIN_SUCCESS,
                  payload: response.data.token
              })
              dispatch({ type: ASYNC_ACTION_END })
              return
          }

          dispatch({
              type: LOGIN_ERROR,
              payload: response
          })
          dispatch({ type: ASYNC_ACTION_END })

      } catch (e) {
          dispatch({
              type: LOGIN_ERROR,
              payload: e
          })
          dispatch({ type: ASYNC_ACTION_END })
      }
  }
)

const clearLoginError = () => ({
    type: CLEAR_LOGIN_ERROR
})

const logOut = () => ({
    type: LOGOUT
})

export { login, logOut, clearLoginError }

