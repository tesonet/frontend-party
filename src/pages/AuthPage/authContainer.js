import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'
import { login, clearLoginError } from './authActions'
import AuthPage from './Components/AuthPage'

const mapDispatchToProps = dispatch => ({
    login: token => dispatch(login(token)),
    clearLoginError: () => dispatch(clearLoginError())
})

export const mapStateToProps = state => ({
    withAuth: state.authReducer.withAuth,
    loginError: state.authReducer.loginError
})

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(AuthPage))