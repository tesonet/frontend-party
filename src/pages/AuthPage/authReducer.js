import { LOGOUT, LOGIN_SUCCESS, LOGIN_ERROR, CLEAR_LOGIN_ERROR } from './authActions'

const token = localStorage.getItem('token')
const isTokenExists = token !== 'null'

const initialState = {
    token: token,
    withAuth: isTokenExists,
    loginError: false
}

export default (state = initialState, action) => {
    switch (action.type) {
        case LOGIN_SUCCESS:
            return {
                ...state,
                token: action.payload,
                withAuth: true
            }

        case LOGIN_ERROR:
            return {
                ...state,
                loginError: true
            }

        case CLEAR_LOGIN_ERROR:
            return {
                ...state,
                loginError: false
            }

        case LOGOUT:
            return {
                ...state,
                token: 'null',
                withAuth: false
            }
        default:
            return state
    }
}