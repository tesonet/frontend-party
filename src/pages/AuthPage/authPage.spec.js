import React from 'react'
import axios from 'axios'
import renderer from 'react-test-renderer'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { mapStateToProps } from './authContainer'
import * as authActions from './authActions'
import { ASYNC_ACTION_END, ASYNC_ACTION_START } from '../../constants/actions'
import loginResponse from '../../mockResponses/loginResponse'
import AuthPage from './Components/AuthPage'

const mockStore = configureStore([thunk])

describe('Authorization page', () => {

    describe('AuthPage container', () => {

        describe('mapStateToProps', () => {

            it('correctly maps state to props', () => {

                const state = {
                    authReducer: {
                        withAuth: false,
                        loginError: false
                    }
                }

                const componentProps = mapStateToProps(state)
                expect(state.authReducer).toEqual(componentProps)
            })
        })

        describe('AuthPage component', () => {
            it('renders correctly', () => {
                const tree = renderer
                  .create(
                    <AuthPage
                      withAuth={false}
                      loginError={false}
                      login={() => {
                      }}
                      clearLoginError={() => {
                      }}
                    />
                  )
                  .toJSON()
                expect(tree).toMatchSnapshot()
            })
        })
    })

    describe('Auth actions', () => {

        describe('Login action', () => {

            let store = null

            beforeEach(() => {
                store = mockStore()
            })

            afterEach(() => {
                store.clearActions()
            })

            it('it should fetch and dispatch server list otherwise dispatch error message', async () => {
                axios.__set({ data: loginResponse, status: 200 })
                await authActions.login()(store.dispatch)
                const actions = store.getActions()

                expect(actions).toContainEqual({ type: ASYNC_ACTION_START })
                expect(actions).toContainEqual({ type: authActions.LOGIN_START })
                expect(actions).toContainEqual({ type: authActions.LOGIN_SUCCESS, payload: loginResponse.token })
                expect(actions).toContainEqual({ type: ASYNC_ACTION_END })
            })

            it('it should dispatch an error action on unsuccessful login', async () => {
                axios.__set({ data: 'Unauthorized', status: 401 })
                await authActions.login()(store.dispatch)
                const actions = store.getActions()

                expect(actions).toContainEqual({ type: ASYNC_ACTION_START })
                expect(actions).toContainEqual({
                    type: authActions.LOGIN_ERROR,
                    payload: { data: 'Unauthorized', status: 401 }
                })
                expect(actions).toContainEqual({ type: ASYNC_ACTION_END })
            })
        })
    })
})