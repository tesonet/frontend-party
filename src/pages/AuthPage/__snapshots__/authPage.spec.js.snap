// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authorization page AuthPage container AuthPage component renders correctly 1`] = `
<div
  className="auth-page"
>
  <div
    className="auth-page_bg"
  />
  <div
    className="auth-page_overlay"
  />
  <div
    className=" auth-form"
  >
    <img
      alt="logo"
      className="logo"
      src="logo.png"
    />
    <form
      className="auth-form_form"
      onSubmit={[Function]}
    >
      <input
        className="auth-form_input user-inp"
        name="username"
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Username"
        type="text"
        value=""
      />
      <input
        className="auth-form_input pass-inp"
        name="password"
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Password"
        type="text"
        value=""
      />
      <button
        className="auth-form_btn btn"
        disabled={true}
        type="submit"
      >
        Log In
      </button>
    </form>
  </div>
</div>
`;
